velox/common/memory/ByteStream.h

```diff
@@ -40,47 +42,58 @@ class OutputStreamListener {
 
 class OutputStream {
  public:
-  explicit OutputStream(
+  explicit OutputStream(OutputStreamListener* listener = nullptr)
+      : listener_(listener) {}
+
+  virtual ~OutputStream() = default;
+
+  virtual void write(const char* s, std::streamsize count) = 0;
+
+  virtual std::streampos tellp() const = 0;
+
+  virtual void seekp(std::streampos pos) = 0;
+
+  OutputStreamListener* listener() const {
+    return listener_;
+  }
+
+ protected:
+  OutputStreamListener* listener_;
+};
+
+class OStreamOutputStream : public OutputStream {
+ public:
+  explicit OStreamOutputStream(
       std::ostream* out,
       OutputStreamListener* listener = nullptr)
-      : out_(out), listener_(listener) {}
+      : OutputStream(listener), out_(out) {}
 
-  void write(const char* s, std::streamsize count) {
+  void write(const char* s, std::streamsize count) override {
     out_->write(s, count);
     if (listener_) {
       listener_->onWrite(s, count);
     }
   }
 
-  std::streampos tellp() const {
+  std::streampos tellp() const override {
     return out_->tellp();
   }
 
-  void seekp(std::streampos pos) {
+  void seekp(std::streampos pos) override {
     out_->seekp(pos);
   }
 
-  OutputStreamListener* listener() const {
-    return listener_;
-  }
-
  private:
   std::ostream* out_;
-  OutputStreamListener* listener_;
 };
 
-template <>
-inline int32_t ByteRange::available<bool>() {
-  return size * 8 - position;
-}
-
 // Stream over a chain of ByteRanges. Provides read, write and
 // comparison for equality between stream contents and memory. Used
 // for streams in repartitioning or for complex variable length data
-// in hash tables.
+// in hash tables. The stream is seekable and supports overwriting of
+// previous content, for example, writing a message body and then
+// seeking to start to write a length header.
 class ByteStream {
```

Comment:
It is super confusing that this class is used as both input and output stream. It would be nice to split it up or document clearly which methods are supposed to be used for reading and which are for writing.